### PR TITLE
sliding window on stream with stride

### PR DESCRIPTION
**Pull request by AlirezaSohofi**
# Description

This PR provides a way to generate sliding windows from a stream of items. It's useful for e.g. training autoregressive models and fixed-window aggregation. It is a generalization of `Composable.batched()`: when `window_size == stride`, the two are equivalent. 

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. 
List any dependencies that are required for this change.

Fixes # issue

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Refactoring including code style reformatting 
- [ ] Other (please describe):

# Checklist:

- [ ] I have read the [contributing guideline doc](https://squirrel-core.readthedocs.io/en/latest/usage/code_of_conduct.html) (external contributors only)
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have kept the PR small so that it can be easily reviewed  
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] All dependency changes have been reflected in the pip requirement files.

### CODE REVIEW

File: squirrel/iterstream/base.py [hunk 1]
```diff
@@ -173,6 +173,40 @@ def batched(
...
+
+        Args:
+            window_size (int): the length of the window
+            deepcopy (bool): If True, each window will be returned as a deepcopy. If items are mutated in the
+                subsequent steps of the pipeline, this should be set to True, otherwise it should be False.
+                Note that deepcopy may incur a substantial cost, so set this parameter carefully.
+            stride (int): the distance that the window moves at each step
+            drop_last_if_not_full (bool): If True, it would only return windows of size `window_size` and drops the
+                last items which have fewer items.
+            min_window_size (int): The minimum length of the window. This argument is only relevant if
```
Comment (SebS94) [inline squirrel/iterstream/base.py:196]: Is this the common way of handling incomplete windows? As a user the more intuitive behaviour for me would be to supply  a default value which is used to complete the last window - then I can filter/ignore it in further processing steps as necessary?

**Review by SebS94 — APPROVED** [review]
LGTM - except for the comment on the expected behaviour.

File: squirrel/iterstream/base.py (hunk 1 above)
Comment (AlirezaSohofi) [reply squirrel/iterstream/base.py:196]: make it possible to fill partial windows with None is a good suggestion, will add this, thanks :+1:

File: squirrel/iterstream/base.py [hunk 2]
```diff
@@ -173,6 +173,40 @@ def batched(
...
+                `drop_last_if_not_full` is set to False, otherwise it's ignored.
```
Comment (pzdkn) [inline squirrel/iterstream/base.py:197]: It is not immediately clear to me  what `min_window_size`  is the minimum window size for. One could add that it is the minimum window size for the last remaining elements.

File: squirrel/iterstream/base.py [hunk 3]
```diff
@@ -317,6 +351,59 @@ def __iter__(self) -> t.Iterator:
...
+    def __iter__(self):
+        it = iter(self.source)
+        _win = []
+
+        while len(_win) < self.window_size:
+            try:
+                _win.append(next(it))
+            except StopIteration:
+                if not self.drop_last_if_not_full:
+                    yield _win
```
Comment (pzdkn) [inline squirrel/iterstream/base.py:381]: I guess we need to yield a deep-copied version here too

Comment (AlirezaSohofi) [reply squirrel/iterstream/base.py:381]: nice catch! Thank you :+1:

File: squirrel/iterstream/base.py (hunk 2 above)
Comment (pzdkn) [reply squirrel/iterstream/base.py:197]: Or even changing name to `min_remaining_items`

**Review by pzdkn — DISMISSED** [review]
LGTM! Thanks Ali!

**Review by pzdkn — APPROVED** [review]
LGTM! Thanks!

**Review by pzdkn — APPROVED** [review]
LGTM!

**Pull request merged**